txsni/snimap.py

```diff
@@ -112,25 +153,48 @@ def __setattr__(self, attr, val):
     def __delattr__(self, attr):
         return delattr(self._obj, attr)
 
-
 @implementer(IOpenSSLServerConnectionCreator)
 class SNIMap(object):
-    def __init__(self, mapping):
+    def __init__(self, mapping, acme_mapping=None):
         self.mapping = mapping
+        self.acme_mapping = acme_mapping
         self._negotiationDataForContext = collections.defaultdict(
             _NegotiationData
         )
         try:
-            self.context = self.mapping['DEFAULT'].getContext()
+            self.context = self.mapping[b'DEFAULT'].getContext()
         except KeyError:
             self.context = CertificateOptions().getContext()
         self.context.set_tlsext_servername_callback(
             self.selectContext
         )
+        self.counter = 0
+
+    def selectAlpn(self, default, connection, protocols):
+        """
+        Trap alpn negotation.
+
+        Acme works by sending a special certificate based on this negotiation.
+        If such a certificate exists in self.acme_mapping, we will respond.
+
+        OpenSSL design flaws prevent us from changing certificates here.
```

Comment:
Can you include a link here to the appropriate ticket so that a new maintainer might know which design flaw?  (maybe the bug will actually be fixed one day).